### PR TITLE
Add throttling to default API

### PR DESCRIPTION
**Pull request by Mr-Groch**
This will add feature requested in #44 
Basically this is copied throttling from entertainment with added bypassing throttling on power state toggle (on/off) and refactored comparing prev and current request data dict

I'm using Ambilight+Hue often, and this is critical for me, because of resource heavy ambilight+hue without throttling

### CODE REVIEW

**Comment by marcelveldt** [comment]
Thanks for the PR! Did you test the change on your setup and does it work smoothly (fix your issue) ?
I think we should move the common logic into a controller so both normal operation and entertainment mode uses same logic.
This can be something for another day though. For now I have one request and that is that the config value for throttling is one generic one so it can be used for both. After all it is a setting for the light, not dependent how/what is calling the light.

**Comment by Mr-Groch** [comment]
Yes, i've tested it with my TV's ambilight+hue few hours, even now it is enabled :) Transitions are smooth, I'm currently testing 500ms throttling and looks good - no so high CPU usage and lag isn't much noticable (I'm using with LED strips on both side of TV in distance about 2m)

You are sure you want to have 1 shared throttling setting? In my opinion it is good for user to use (or not) less throttling for standard API (for most this is only for sdandard light controlling) and higher throttling for much heavier entertainment mode...

File: emulated_hue/api.py [hunk 1]
```diff
@@ -627,14 +642,50 @@ async def __async_light_action(self, entity: dict, request_data: dict) -> None:
...
+        """Minimalistic form of throttling, only allow updates to a light within a timespan."""
+
+        # check if data changed
+        # when not using udp no need to send same light command again
+        prev_data = self._prev_data.get(light_id, {})
+
+        # force to update if power state changed
+        if prev_data.get(const.HASS_STATE_ON, True) != light_data.get(const.HASS_STATE_ON, True):
+            return True
+        if prev_data.get(const.HUE_ATTR_BRI, 0) == light_data.get(const.HUE_ATTR_BRI, 0)\
```
Comment (marcelveldt) [inline emulated_hue/api.py]: remove the backslash and put it all in brackets. black formatter will do this automagically for you btw.

**Comment by marcelveldt** [comment]
> Yes, i've tested it with my TV's ambilight+hue few hours, even now it is enabled :) Transitions are smooth, I'm currently testing 500ms throttling and looks good - no so high CPU usage and lag isn't much noticable (I'm using with LED strips on both side of TV in distance about 2m)
> 
> You are sure you want to have 1 shared throttling setting? In my opinion it is good for user to use (or not) less throttling for standard API (for most this is only for sdandard light controlling) and higher throttling for much heavier entertainment mode...

Throttling is all about not sending too many commands per interval to a light so it applies to both usecases.
So one throttle setting for each light to override the default of 0 (no throttling), regardless of how the command is issued.

**Comment by alexyao2015** [comment]
I agree, there isn't really a reason to have different throttling variables given that there is no difference between how we handle sending requests to HA through either the standard API or the entertainment API. The only difference is that entertainment will send more packets.

**Comment by alexyao2015** [comment]
As a sidenote, I think we might want to look into refactoring the api so that it's broken apart into different components like the hue documentation. Ultimately, the api is large and handling all of it in a single file seems hard to follow.

**Comment by marcelveldt** [comment]
> As a sidenote, I think we might want to look into refactoring the api so that it's broken apart into different components like the hue documentation. Ultimately, the api is large and handling all of it in a single file seems hard to follow.

That's what I mean with refactoring the logic into one or more controllers and let the api just be the api.
I'll put it on the TODO list.

**Comment by Mr-Groch** [comment]
Changed to only one throttling config variable

**Comment by marcelveldt** [comment]
Thanks!

**Comment by alexyao2015** [comment]
Something fundamentally broke after this PR. None of the lights respond anymore. @Mr-Groch Can you do some more testing to figure out what happened? I'll need to revert this PR in the meantime.

**Comment by Mr-Groch** [comment]
> Something fundamentally broke after this PR. None of the lights respond anymore. @Mr-Groch Can you do some more testing to figure out what happened? I'll need to revert this PR in the meantime.

Can you please descibe what's happening?

**Comment by Mr-Groch** [comment]
OK, I think I see it now. Initial power state change does not work without additional data (color or color temperature), because new and old atrributes are empty. This isn't happening when you initailyy turn on with color or colortemperature attribute. I will fix that

**Comment by alexyao2015** [comment]
We also would need to make sure that it respects the default setting of no throttling. Even with that issue, at 0 throttling, it shouldn't have caused that to happen.

**Comment by alexyao2015** [comment]
I would also like to remind you that you can test your own repository's forks using the dev addon as well. View the documentation within HASS for the format to use.
